ddl/table.go

```diff
@@ -948,56 +956,191 @@ func updateVersionAndTableInfo(t *meta.Meta, job *model.Job, tblInfo *model.Tabl
+		}
 	}
+	return
+}
 
-	err = checkAddPartitionNameUnique(tblInfo, partInfo)
-	if err != nil {
-		job.State = model.JobStateCancelled
-		return ver, errors.Trace(err)
+func storeHasEngineTiFlashLabel(store *metapb.Store) bool {
+	for _, label := range store.Labels {
+		if label.Key == "engine" && label.Value == "tiflash" {
+			return true
+		}
 	}
+	return false
+}
 
-	updatePartitionInfo(partInfo, tblInfo)
-	ver, err = updateVersionAndTableInfo(t, job, tblInfo, true)
+// TODO: It may have the issue when two clients concurrently add partitions to a table.
+func onAddTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error) {
+	// Handle the rolling back job
+	if job.IsRollingback() {
+		ver, err := onDropTablePartition(t, job)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		return ver, nil
+	}
+
+	tblInfo, partInfo, partDefPointers, err := checkAddPartition(t, job)
 	if err != nil {
-		return ver, errors.Trace(err)
+		return ver, err
 	}
-	// Finish this job.
-	job.FinishTableJob(model.JobStateDone, model.StatePublic, ver, tblInfo)
-	asyncNotifyEvent(d, &util.Event{Tp: model.ActionAddTablePartition, TableInfo: tblInfo, PartInfo: partInfo})
+
+	// partDefPointers's len is 0 meaning the job is in the initial state of add partition.
+	// Here should use partInfo from job directly and do some check action.
+	if len(partDefPointers) == 0 {
+		err = checkAddPartitionTooManyPartitions(uint64(len(tblInfo.Partition.Definitions) + len(partInfo.Definitions)))
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return ver, errors.Trace(err)
+		}
+
+		err = checkAddPartitionValue(tblInfo, partInfo)
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return ver, errors.Trace(err)
+		}
+
+		err = checkAddPartitionNameUnique(tblInfo, partInfo)
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return ver, errors.Trace(err)
+		}
+		partDefPointers = updatePartitionInfo(partInfo, tblInfo)
+	}
+
+	originalState := partDefPointers[0].State
+	switch partDefPointers[0].State {
+	case model.StateNone:
+		// none -> delete only
+		job.SchemaState = model.StateDeleteOnly
+		for _, pdDef := range partDefPointers {
+			pdDef.State = model.StateDeleteOnly
+		}
+		ver, err = updateVersionAndTableInfoWithCheck(t, job, tblInfo, originalState != partDefPointers[0].State)
+	case model.StateDeleteOnly:
+		// delete only -> public
+		// Here need do some tiflash replica complement check.
+		if tblInfo.TiFlashReplica != nil && tblInfo.TiFlashReplica.Available {
+			// For available state, the new added partition should wait it's replica to
+			// be finished. Otherwise the query to this partition will be blocked.
+			needWait, err := checkPartitionReplica(partDefPointers, d, job)
+			if err != nil {
+				return ver, err
+			}
+			if needWait {
+				// The new added partition hasn't been replicated.
+				// Do nothing to the job this time, wait next worker round.
+				time.Sleep(tiFlashCheckTiDBHttpAPIInterval / 2)
```

Comment:
If we need to do `tiFlashCheckTiDBHttpAPIInterval / 2`, why not set `tiFlashCheckTiDBHttpAPIInterval=2500ms` directly?